File: packages/knowledge-sdk/src/mcp.ts
#!/usr/bin/env node
import readline from "node:readline";
import { CumulusKnowledge } from "./index.js";

const root = process.env.CUMULUS_ROOT ?? process.cwd();
const client = new CumulusKnowledge({ root });

const tools = [
  { name: "graph_view", description: "Return a semantic graph view with readable labels, legend, layout, filters, and evidence.", inputSchema: { type: "object", properties: { preset: { type: "string" } } } },
  { name: "search", description: "Search indexed nodes and chunks.", inputSchema: { type: "object", properties: { query: { type: "string" }, limit: { type: "integer" } }, required: ["query"] } },
  { name: "fetch", description: "Fetch a node by ID.", inputSchema: { type: "object", properties: { id: { type: "string" } }, required: ["id"] } },
  { name: "expand_neighbors", description: "Expand graph neighbors.", inputSchema: { type: "object", properties: { id: { type: "string" }, depth: { type: "integer" } }, required: ["id"] } },
  { name: "find_paths", description: "Find a graph path.", inputSchema: { type: "object", properties: { from_id: { type: "string" }, to_id: { type: "string" } }, required: ["from_id", "to_id"] } },
  { name: "index_status", description: "Return local index status.", inputSchema: { type: "object", properties: {} } },
  { name: "ingest", description: "Index the configured root.", inputSchema: { type: "object", properties: { profile: { type: "string" } } } },
  { name: "source_trace", description: "Return source evidence for a semantic graph node.", inputSchema: { type: "object", properties: { id: { type: "string" }, preset: { type: "string" } }, required: ["id"] } },
];

const rl = readline.createInterface({ input: process.stdin, terminal: false });

rl.on("line", async (line) => {
  if (!line.trim()) return;
  let request: any;
  try {
    request = JSON.parse(line);
  } catch (error) {
    write({ jsonrpc: "2.0", id: null, error: { code: -32700, message: String(error) } });
    return;
  }
  const id = request.id;
  if (id === undefined) return;
  try {
    const result = await handle(request.method, request.params ?? {});
    write({ jsonrpc: "2.0", id, result });
  } catch (error) {
    write({ jsonrpc: "2.0", id, error: { code: -32603, message: error instanceof Error ? error.message : String(error) } });
  }
});

async function handle(method: string, params: any): Promise<unknown> {
  if (method === "initialize") {
    return {
      protocolVersion: "2025-06-18",
      serverInfo: { name: "cmls-knowledge", version: "0.1.0" },
      capabilities: { tools: {}, resources: {}, prompts: {} },
    };
  }
  if (method === "tools/list") return { tools };
  if (method === "resources/list") {
    return { resources: [
      { uri: "cumulus://snapshot/current", name: "Current Graph Snapshot", mimeType: "application/json" },
      { uri: "cumulus://graph-view/current", name: "Current Semantic Graph View", mimeType: "application/json" },
    ] };
  }
  if (method !== "tools/call") return {};

  const name = params.name;
  const args = params.arguments ?? {};
  if (name === "graph_view") return toolResult((await client.getGraphView(args.preset ?? "full")).data);
  if (name === "search") return toolResult((await client.query(args.query, { limit: args.limit })).data);
  if (name === "fetch") return toolResult((await client.getNode(args.id)).data);
  if (name === "expand_neighbors") return toolResult((await client.expandNeighbors(args.id, args.depth ?? 1)).data);
  if (name === "find_paths") return toolResult((await client.findPaths(args.from_id, args.to_id, args.max_depth ?? 6)).data);
  if (name === "index_status") return toolResult((await client.indexStatus()).data);
  if (name === "ingest") return toolResult((await client.index({ profile: args.profile ?? "all" })).data);
  if (name === "source_trace") {
    const view = (await client.getGraphView(args.preset ?? "full")).data;
    return toolResult({ node_id: args.id, evidence: view.evidence.filter((item) => item.node_id === args.id) });
  }
  throw new Error(`unknown tool: ${name}`);
}

function toolResult(data: unknown): unknown {
  return {
    content: [{ type: "text", text: JSON.stringify(data, null, 2) }],
    structuredContent: data,
  };
}

function write(value: unknown): void {
  process.stdout.write(`${JSON.stringify(value)}\n`);
}

File: crates/cumulus-knowledge-core/src/mcp.rs
use serde_json::{json, Value};

pub fn tool_definitions() -> Vec<Value> {
    vec![
        json!({
            "name": "graph_view",
            "title": "Cumulus Graph View",
            "description": "Return a semantic presentation graph with readable display labels, legend, layout, filters, and evidence links.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "preset": {"type": "string", "enum": ["source", "finance", "timeline", "risk", "full"], "default": "full"},
                    "limit": {"type": "integer", "default": 800}
                }
            },
            "annotations": {"readOnlyHint": true, "destructiveHint": false, "openWorldHint": false}
        }),
        json!({
            "name": "search",
            "title": "Search Cumulus Knowledge",
            "description": "Search indexed nodes and chunks. Returns compact hits with IDs, snippets, scores, and cumulus:// resource URIs.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "query": {"type": "string"},
                    "limit": {"type": "integer", "default": 10}
                },
                "required": ["query"]
            },
            "annotations": {"readOnlyHint": true, "destructiveHint": false, "openWorldHint": false}
        }),
        json!({
            "name": "fetch",
            "title": "Fetch Cumulus Node Or Chunk",
            "description": "Fetch one node or chunk by stable ID.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "id": {"type": "string"}
                },
                "required": ["id"]
            },
            "annotations": {"readOnlyHint": true, "destructiveHint": false, "openWorldHint": false}
        }),
        json!({
            "name": "expand_neighbors",
            "title": "Expand Cumulus Neighbors",
            "description": "Return the local graph neighborhood for a node.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "id": {"type": "string"},
                    "depth": {"type": "integer", "default": 1}
                },
                "required": ["id"]
            },
            "annotations": {"readOnlyHint": true, "destructiveHint": false, "openWorldHint": false}
        }),
        json!({
            "name": "find_paths",
            "title": "Find Cumulus Path",
            "description": "Find a short graph path between two node IDs.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "from_id": {"type": "string"},
                    "to_id": {"type": "string"},
                    "max_depth": {"type": "integer", "default": 6}
                },
                "required": ["from_id", "to_id"]
            },
            "annotations": {"readOnlyHint": true, "destructiveHint": false, "openWorldHint": false}
        }),
        json!({
            "name": "index_status",
            "title": "Cumulus Index Status",
            "description": "Return counts and local index health.",
            "inputSchema": {"type": "object", "properties": {}},
            "annotations": {"readOnlyHint": true, "destructiveHint": false, "openWorldHint": false}
        }),
        json!({
            "name": "ingest",
            "title": "Ingest Cumulus Folder",
            "description": "Index the configured folder into the local Cumulus graph.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "profile": {"type": "string", "enum": ["code", "docs", "facility", "all"], "default": "all"}
                }
            },
            "annotations": {"readOnlyHint": false, "destructiveHint": false, "openWorldHint": false}
        }),
        json!({
            "name": "summarize_subgraph",
            "title": "Summarize Cumulus Subgraph",
            "description": "Return a compact text summary of a local graph neighborhood.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "id": {"type": "string"},
                    "depth": {"type": "integer", "default": 1}
                },
                "required": ["id"]
            },
            "annotations": {"readOnlyHint": true, "destructiveHint": false, "openWorldHint": false}
        }),
        json!({
            "name": "source_trace",
            "title": "Trace Cumulus Source Evidence",
            "description": "Return source/evidence links for a semantic graph node.",
            "inputSchema": {
                "type": "object",
                "properties": {
                    "id": {"type": "string"},
                    "preset": {"type": "string", "enum": ["source", "finance", "timeline", "risk", "full"], "default": "full"}
                },
                "required": ["id"]
            },
            "annotations": {"readOnlyHint": true, "destructiveHint": false, "openWorldHint": false}
        }),
    ]
}

pub fn prompt_definitions() -> Vec<Value> {
    vec![
        json!({"name": "summarize_neighborhood", "title": "Summarize Neighborhood", "description": "Summarize the important nodes and relationships around a Cumulus node."}),
        json!({"name": "trace_decision", "title": "Trace Decision", "description": "Trace a decision through source files, docs, and linked graph nodes."}),
        json!({"name": "explain_symbol", "title": "Explain Symbol", "description": "Explain a code symbol with nearby context and citations."}),
        json!({"name": "audit_context", "title": "Audit Context", "description": "Check whether retrieved context is complete, concise, and grounded."}),
    ]
}
